ckanext/googleanalytics/plugin/pylons_plugin.py

```diff
@@ -0,0 +1,181 @@
+# -*- coding: utf-8 -*-
+import Queue as queue
+
+import hashlib
+import importlib
+
+import ckan.plugins as plugins
+import ckan.plugins.toolkit as tk
+
+from ckan.controllers.package import PackageController
+from pylons import config
+from routes.mapper import SubMapper
+
+
+class GAMixinPlugin(plugins.SingletonPlugin):
+    plugins.implements(plugins.IRoutes)
+
+    analytics_queue = queue.Queue()
+
+    def before_map(self, map):
+        """Add new routes that this extension's controllers handle.
+
+        See IRoutes.
+
+        """
+        # Helpers to reduce code clutter
+        GET = dict(method=["GET"])
+        PUT = dict(method=["PUT"])
+        POST = dict(method=["POST"])
+        DELETE = dict(method=["DELETE"])
+        GET_POST = dict(method=["GET", "POST"])
+        # intercept API calls that we want to capture analytics on
+        register_list = [
+            "package",
+            "dataset",
+            "resource",
+            "tag",
+            "group",
+            "related",
+            "revision",
+            "licenses",
+            "rating",
+            "user",
+            "activity",
+        ]
+        register_list_str = "|".join(register_list)
+        # /api ver 3 or none
+        with SubMapper(
+            map,
+            controller="ckanext.googleanalytics.controller:GAApiController",
+            path_prefix="/api{ver:/3|}",
+            ver="/3",
+        ) as m:
+            m.connect(
+                "/action/{logic_function}",
+                action="action",
+                conditions=GET_POST,
+            )
+
+        # /api ver 1, 2, 3 or none
+        with SubMapper(
+            map,
+            controller="ckanext.googleanalytics.controller:GAApiController",
+            path_prefix="/api{ver:/1|/2|/3|}",
+            ver="/1",
+        ) as m:
+            m.connect("/search/{register}", action="search")
+
+        # /api/rest ver 1, 2 or none
+        with SubMapper(
+            map,
+            controller="ckanext.googleanalytics.controller:GAApiController",
+            path_prefix="/api{ver:/1|/2|}",
+            ver="/1",
+            requirements=dict(register=register_list_str),
+        ) as m:
+
+            m.connect("/rest/{register}", action="list", conditions=GET)
+            m.connect("/rest/{register}", action="create", conditions=POST)
+            m.connect("/rest/{register}/{id}", action="show", conditions=GET)
+            m.connect("/rest/{register}/{id}", action="update", conditions=PUT)
+            m.connect(
+                "/rest/{register}/{id}", action="update", conditions=POST
+            )
+            m.connect(
+                "/rest/{register}/{id}", action="delete", conditions=DELETE
+            )
+
```

Comment:
The routes for `GAOrganizationController` and `GAPackageController` to catch the page views for packages in organizations. A view for a package is also counted as a view for the associated org.

https://github.com/OpenGov-OpenData/ckanext-googleanalytics/blob/97d142a202149817c2c3d91eb5418903255f4d8f/ckanext/googleanalytics/plugin.py#L245-L253